src/sage/rings/species.py

```diff
@@ -0,0 +1,2255 @@
+    symmetric group.
+
+    Two conjugacy classes of subgroups are equal if they have the
+    same degree (say `n`) and are conjugate within `S_n`.
+    """
+    @staticmethod
+    def _invariant(C):
+        r"""
+        Return an invariant of the permutation group ``C`` used for
+        hashing and as key in the cache.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import ConjugacyClassOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassOfDirectlyIndecomposableSubgroups
+            sage: G = PermutationGroup([[(1,2),(3,4)]])
+            sage: H = PermutationGroup([[(1,3),(2,4)], [(1,2),(3,4)]])
+            sage: K = PermutationGroup([[(1,2,3,4)]])
+            sage: C._invariant(G)
+            (2, (2, 2))
+            sage: C._invariant(H)
+            (4, (4,))
+            sage: C._invariant(K)
+            (4, (4,))
+        """
+        return C.cardinality(), tuple(sorted(len(o) for o in C.orbits()))
+
+    @staticmethod
+    def __classcall__(cls, parent, C):
+        """
+        Normalize the input for unique representation.
+
+        TESTS::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
+
+        Check that the domain is standardized::
+
+            sage: G = PermutationGroup([("a", "b", "c"), ("d", "a"), ("d", "b"), ("d", "c")])
+            sage: C(G)
+            ((2,4,3), (1,2))
+
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: H = PermutationGroup([[(1,2,3,4),(5,6),(7,8),(9,10)]]); H
+            Permutation Group with generators [(1,2,3,4)(5,6)(7,8)(9,10)]
+            sage: C(G) is C(H)
+            True
+
+        Two different transitive groups with the same cardinality::
+
+            sage: a = C(PermutationGroup([[(1,3),(2,4)], [(1,4),(2,3)]]))
+            sage: b = C(PermutationGroup([[(1,3,2,4)]]))
+            sage: a == b
+            False
+        """
+        def is_equal(elm):
+            return SymmetricGroup(elm._C.degree()).are_conjugate(elm._C, C)
+
+        def standardize(C):
+            if not C.degree():
+                return SymmetricGroup(0)
+            # an directly indecomposable group is transitive, so we
+            # can use the gap group without worrying about the domain
+            G = C.gap()
+            sorted_orbits = sorted(G.Orbits().sage(), key=len, reverse=True)
+            pi = PermutationGroupElement([e for o in sorted_orbits for e in o])
+            G = PermutationGroup(G.SmallGeneratingSet().sage())
+            return G.conjugate(pi.inverse())
+
+        key = cls._invariant(C)
+        if key in parent._cache:
+            lookup = parent._cache[key]
+            for elm in lookup:
+                if is_equal(elm):
```

Comment:
Micro-optimization: Wouldn't it be better to just inline the `is_equal()` function since it is only used here? Plus, it might be better to store (locally) in a `dict` the `SymmetricGroup`s you construct to improve the lookup speed. Nothing needs to be done here though.

Reply:
The local `is_equal` is to separate the caching from the rest, and this structure appears another time in `AtomicSpeciesElement`.  I think it is OK the way it is, but I don't mind to change it either.

The `SymmetricGroup`s are cached globally - they have `CachedRepresentation`.

Reply:
I don't really follow your logic, but okay. In `AtomicSpeciesElement`, there is more going on, so it makes sense to separate it out there. Yes, the are cached, but the lookup time for those symmetric groups are far slower than getting it from a `dict`:
```
sage: S5 = SymmetricGroup(5)
sage: %timeit SymmetricGroup(5)
2.45 µs ± 18 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
sage: d = {5: S5}
sage: %timeit d[5]
47.5 ns ± 0.46 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
This could add up a bit when checking against a lot of elements. I just don't know how common that is. Well, in either case, I expect it to be a mirco-optimization. As such, we can leave it.